File: ml-benchmark/synduce/list/last.f
Inductive List = elt Int | cons {Int, List};
Inductive CList = single Int | concat {CList, CList};

spec = fix (
  \f: List -> Int. \x: List.
  match x with
    elt a -> a
  | cons {h, t} -> f t
  end
);

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    elt a -> cons {a, b}
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
    \f: CList -> Compress List. \xs: CList.
    match xs with
      single a -> elt a
    | concat {a, b} -> cat (f a) (f b)
    end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/last.ml
type list = Elt of int | Cons of int * list
type clist = Single of int | Concat of clist * clist

let rec spec x =
  match x with
  | Elt a -> a
  | Cons (h, t) -> spec t

let rec cat a b =
  match a with
  | Elt a0 -> Cons (a0, b)
  | Cons (h, t) -> Cons (h, (cat t b))

val repr: clist -> list compress
let rec repr xs =
  match xs with
  | Single a -> Elt a
  | Concat (a, b) -> cat (repr a) (repr b)

let program x = spec (repr x)